### ISSUE TITLE
relationchoice not working ...

### ISSUE THREAD

**Issue by iham** [collective/collective.z3cform.datagridfield#67]
```python
from plone.app.vocabularies.catalog import CatalogSource
from z3c.relationfield.schema import RelationChoice
from zope import schema
from zope.interface import Interface

from collective.z3cform.datagridfield import DataGridFieldFactory
from collective.z3cform.datagridfield import DictRow
from plone.autoform import directives
from plone.supermodel import model

class IItemRow(Interface):
    note = schema.TextLine(
        title=u'Note',
        required=True,
        default=u'',
    )
    events = RelationChoice(
        title=u'Event',
        source=CatalogSource(portal_type=['Event']),
        required=True,
    )

class IEventsGroup(model.Schema):
    directives.widget(employees=DataGridFieldFactory)
    employees = schema.List(
        title=u'A list of events with notes.',
        required=False,
        value_type=DictRow(
            title=u'Event',
            schema=IEventRow,
        )
    )
```

IEventsGroup is a globally addable contenttype.

the widget gets displayed, but when clicking in the relationchoice widget i get a spinner and no result. calling the view "getSource" i get **{"error": "Vocabulary lookup not allowed."}**
this seems to occur as self.context.context has **NO_VALUE**

any help?

**Comment by thet** [comment]
If you want to make use of the related items widget, I'd try not to use a CatalogSource but the following:

```
# -*- coding: utf-8 -*-
from collective.z3cform.datagridfield import DataGridFieldFactory
from collective.z3cform.datagridfield import DictRow
from plone.app.z3cform.widget import RelatedItemsFieldWidget
from plone.autoform import directives
from plone.supermodel import model
from z3c.relationfield.schema import RelationChoice
from zope import schema
from zope.interface import Interface


class IEventRow(Interface):
    note = schema.TextLine(
        title=u'Note',
        required=True,
        default=u'',
    )

    directives.widget(
        'events',
        RelatedItemsFieldWidget,
        pattern_options={
            'selectableTypes': ['Event'],
        }
    )
    events = RelationChoice(
        title=u'Event',
        vocabulary='plone.app.vocabularies.Catalog',
        required=True,
    )


class IEventsGroup(model.Schema):
    directives.widget(employees=DataGridFieldFactory)
    employees = schema.List(
        title=u'A list of events with notes.',
        required=False,
        value_type=DictRow(
            title=u'Event',
            schema=IEventRow,
        )
    )
```

**Comment by iham** [comment]
i will give that a try! thx @thet

**Comment by iham** [comment]
```
Traceback (innermost last):
  Module ZPublisher.Publish, line 138, in publish
  Module ZPublisher.mapply, line 77, in mapply
  Module ZPublisher.Publish, line 48, in call_object
  Module plone.z3cform.layout, line 66, in __call__
  Module plone.z3cform.layout, line 50, in update
  Module plone.dexterity.browser.edit, line 58, in update
  Module plone.z3cform.fieldsets.extensible, line 59, in update
  Module plone.z3cform.patch, line 30, in GroupForm_update
  Module z3c.form.group, line 132, in update
  Module z3c.form.form, line 136, in updateWidgets
  Module z3c.form.field, line 277, in update
  Module z3c.form.browser.multi, line 63, in update
  Module z3c.form.browser.widget, line 171, in update
  Module z3c.form.widget, line 496, in update
  Module Products.CMFPlone.patches.z3c_form, line 46, in _wrapped
  Module z3c.form.widget, line 132, in update
  Module z3c.form.widget, line 491, in value
  Module collective.z3cform.datagridfield.datagridfield, line 153, in updateWidgets
  Module collective.z3cform.datagridfield.datagridfield, line 127, in getWidget
  Module z3c.form.browser.widget, line 171, in update
  Module z3c.form.object, line 217, in update
  Module z3c.form.object, line 208, in updateWidgets
  Module collective.z3cform.datagridfield.autoform, line 56, in update
  Module plone.z3cform.fieldsets.extensible, line 59, in update
  Module plone.z3cform.patch, line 30, in GroupForm_update
  Module z3c.form.group, line 132, in update
  Module collective.z3cform.datagridfield.datagridfield, line 315, in updateWidgets
  Module collective.z3cform.datagridfield.autoform, line 40, in updateWidgets
  Module z3c.form.form, line 136, in updateWidgets
  Module z3c.form.field, line 277, in update
  Module plone.app.z3cform.widget, line 400, in update
  Module z3c.form.browser.text, line 36, in update
  Module z3c.form.browser.widget, line 171, in update
  Module Products.CMFPlone.patches.z3c_form, line 46, in _wrapped
  Module z3c.form.widget, line 112, in update
  Module zope.schema._field, line 312, in bind
  Module Zope2.App.schema, line 33, in get
  Module plone.app.vocabularies.catalog, line 627, in __call__
AttributeError: 'NO_VALUE' object has no attribute 'getPhysicalPath'
```

**Comment by iham** [comment]
folgenden "fix" hab ich jetzt gemacht, weil der context immer ein problem macht:

CatalogVocabularyFactory - overrides.zcml
und den __call__() so umgebogen, dass der context portal ist, falls er NO_VALUE oder dict ist.

das geht jetzt. bin aber nicht sehr happy da ein basis vocab umbiegen zu müssen.

ich betrachte das also noch nicht als gelöst.

**Comment by pbauer** [comment]
@iham did you get this working? I ran into the same problem today using the current master that includes https://github.com/collective/collective.z3cform.datagridfield/pull/71).

**Comment by iham** [comment]
@pbauer sadly: no. even after hours of desperation and agony ;)

i guess this is broken by concept as complex fields using ```context``` don't work. there is no context, so no way to get a catalog, a portal_root or anything at all.

i'd say: **YAFOWIL** as @rnixx and @jensens are working on that and more and more fields/widgets/forms and types start looking good and useable. this will allow all sorts of complex fields (compounds) and sorting works like a charm (better than z3c.form) and no change in how to write behaviors, forms, ...

**Comment by laulaz** [comment]
Fixed by #150  👍
